### DISCUSSION TITLE
General Advice: Saving document changes

### DISCUSSION THREAD

**Discussion by sgthuth** [pymupdf/PyMuPDF#2239] [General]
I've been trying to read through PyMuPDF documentation and the examples don't quite fit my situation.  What would be the best way or most efficient way to save changes to my documents?

What I'm trying to do is sift through a plethura of PDF files, scan each page of them for certain criteria.  Based on the criteria found or not on each page, I may need to remove text, may need to cripple/image the page, and/or may need to add text.  Below is a basic rundown.  My problem is there are instances where pages are blank or missing the added text... 

How / Where would the best place be to place a save?  Looking at this now... I'm thinking I may need a temporary file... remove text if applicable from current page, save as a temp.pdf, image specified page in temp.pdf, then add my text to the page in the temp.pdf, then save that as my resulting output.

```python
import fitz
import glob
from shutil import copy2

def AlterPDF():
    SrcPath = 'SomePathA'
    DstPath = 'SomePathB'
    FLS = glob.glob(SrcPath+'*.pdf')
    for f in FLS:
        fn = f.split('\\')[-1]
        SRC = fitz.open(f)
        DST = fitz.open()
        RDTs = []
        ADTs = []
        CHGs = 0
        x = 1
        y = 1
        CripplePDF = False
        nbrOfPages = len(SRC)
        for p in range(nbrOfPages):
             page = SRC[p]
             rect = page.bound()

             # Tests, the 3 primary needs are below
             fndTxt = page.search_for('SomeArbitraryText')
             if len(fndTxt) > 0:
                 RDTs.append(['SomeArbitraryText', fndTxt[0]])
             fndTxt = page.search_for('SomeArbitraryText')
             if len(fndTxt) > 0:
                 CripplePDF = True
             fndTxt = page.search_for('SomeArbitraryText')
             if len(fndTxt) > 0:
                 ADTs.append(['SomeArbitraryText', fitz.Point(x, y)])

             # Alterations
             for r in RDTs:
                 # Remove text from current page
                 page.add_redact_annot(r[1])
                 page.apply_redactions()
                 page.apply_redactions(images=fitz.PDF_REDACT_IMAGE_REMOVE)
                 CHGs += 1
             if CripplePDF:
                 if PageHasContent:
                     # Turn current page into an image
                     pix = page.get_pixmap(dpi=600, colorspace=fitz.csGRAY)
                     jpeg = pix.pil_tobytes(format='JPEG', mode='1', dpi=[450, 450], quality=25, optimize=true)
                     newPage = DST.new_page(width=rect.width, height=rect.height)
                     newPage.insert_image(rect, stream=jpeg, overlay=False)
                 else:
                     # Insert Blank Page
                     newPage = DST.new_page(width=rect.width, height=rect.height)
                 CHGs += 1
             else:
                # Keep Page As-Is
                DST.insert_pdf(SRC, from_page=p, to_page=p)
            for a in ADTs:
                # Add text to page
                DST[p].insert_text(a[1], a[0], color=(0.0, 0.0, 0.0), fontsize=7, fontname='Helv')
                CHGs += 1
        if CHGs > 0:
            DST.save(DstPath+fn , garbage=3, deflate=True)
        else:
            copy2(f, DstPath+fn)
```

**Comment by JorjMcKie** [comment]
As you mentioned, it looks like using temporary files or folders is the best alternative.
You don't seem to need PyMuPDF-specific advice as such - chapeau!
I briefly considered using incremental saves (i.e. directly to the original file), but that would incur a high risk of incomplete / undefined situations in case of errors and also potentially large increases of file sizes when addind / replacing images.
The only concern would be how to ensure maintaining temporary storage long enough (and beyond any exceptional situations), so you finally can copy it to permanent storage.

**Reply by sgthuth** [reply]
Thanks @JorjMcKie,

As always... Much help.  I starting going down the temporary file route.  Once I get this coded and tested... I'll put a snip on this discussion.  I don't tend to run into exceptionally large file sizes in succession... Usually many files with a max of 25 pages each... normally they're about 5 pages... sometimes the sum of all the files is gigs but I haven't ran into any issues with processing using PyMuPDF... what is chapeau?  There may come a time where I get a PDF file that's a combined version of all those little files... thanks for bringing that up... I will need to code to split the files up... probably the simplest way to address memory.

**Reply by JorjMcKie** [reply]
Thank you! 
> *chapeau!* = French for "hats off"!

**Reply by sgthuth** [reply]
Haha... You said it seems I don't need PyMuPDF... if I'm in python primarily, what would I use instead... I haven't had the extent of success with PyPDF2 or other PDF libraries.  As far as incrementally saving... if I'm editing a file in this way, is it wise/beneficial to save after each change (as I'm looping through the pages) or is better to save as I've got currently listed above (once I've gone through all the pages / edits)?

**Reply by JorjMcKie** [reply]
I didn't say you don't need PyMuPDF! On the contrary: you couldn't do without it!
What I intended to say: you don't seem to need (much) help to use it professionally.

Anyway, as per incremental saves:
This is an extremely fast way to store information, because the original file is not rewritten, but any changes you make are being appended to the file. You would not need to fiddle around with temp storage, etc.
It is generally advisable to execute the save once you are done with the file - and not save after every change.

But the original file will be changed of course - there is no rollback in case anything goes wrong midway. Also, any old data being changed are physically still there in the file. When the updated file is displayed / printed they are ignored / invisible -- but in terms of data protection / confidentiality they can still be accessed.
